Round robin validation tests

qualand opened the pull request:
Created validation tests for the native runner using the Heliostat CST template. These results are based on the round robin study published here: https://doi.org/10.1016/j.solener.2025.113785

This closes #78 for the native runner. Other runners can use this framework to perform the same validations cases.

qualand committed:
Add separate back optics to Heliostat class

Introduces a new OpticalProperties member for the back surface of Heliostat and methods to set mirror and back optics independently. Updates usage in tests and simulation logic to use set_mirror_optics instead of set_optics, and ensures back optics are applied correctly in element construction.

qualand committed:
adding off-axis canting

initially adding, test is passing with low tolerances. Will revisit.

qualand committed:
Refactor heliostat z-rotation calculation and add test

Updated the z-rotation calculation in heliostat.cpp. Added a comprehensive unit test for a single facet flat heliostat oriented north of receiver, verifying geometry, ray tracing, and power absorption with sunshape and surface errors.

qualand committed:
Fixed python heliostat template test

qualand committed:
Add single heliostat tests task 1a and 1b of round robin results

Replaced manual azimuth calculation in heliostat.cpp with atan2 for improved accuracy. Updated tracing_errors.cpp to consistently convert errors from mrad to rad and removed redundant conversions. Moved the SingleFacetFlat_North test from heliostat_test.cpp to a new single_heliostat_test.cpp file, expanding it with additional focused and southeast position tests. Updated CMakeLists.txt to include the new test file. Added a flux binning method for the test environment. This should be standardized.

qualand committed:
Adding the remaining Phase 1 cases

qualand committed:
Fix out-of-bounds access in st_tree_node::m_proc

Added a bounds check before accessing key.at(index) to prevent out-of-range exceptions. The method now returns 'this' if the index is out of bounds.

qualand committed:
adding round robin study files to the repo

cmake copies them to the binary build directory.

qualand committed:
updated  single heliostat tests to use SimulationResult

qualand committed:
Add heliostat field simulation unit test

Introduces heliostat_field_test.cpp to test simulation of a heliostat field, including ray tracing, receiver flux mapping, and validation of power and flux results. Most tests are commented out currently. Updates CMakeLists.txt to include the new test.

qualand committed:
Add advanced heliostat field simulation tests

Expanded the heliostat field simulation test suite to cover multiple configurations, including slant-focused and band-focused single and multi-facet heliostats, center and scatter aim strategies, and different solar positions. Added methods for reading expected results, checking simulation outputs, and saving flux maps for validation. Also added a get_area() method to the Heliostat class for area calculations.

qualand committed:
added remaining full field tests

commented out because native runner is too slow to run as tests.

qualand committed:
Merge branch 'develop' into round_robin_validation_tests

qualand committed:
Reset SunRayCount

Added initialization of System->SunRayCount to 0 to ensure correct ray counting at the beginning of each run.

qualand committed:
Refactor heliostat field tests for modularity and reduce duplication

Consolidated and refactored heliostat field test cases to reduce code duplication and improve maintainability. Introduced helper methods for common simulation steps, parameterized test logic for different configurations, and replaced repetitive test cases with more modular and reusable code blocks. Also added TODOs for further refactoring in single heliostat tests.

qualand committed:
Remove unused print_info and print_sun_info parameters

Eliminated the print_info and print_sun_info parameters from several helper methods in heliostat_field_test.cpp, as they were no longer needed. Updated all method calls accordingly to improve code clarity.

qualand committed:
Refactor and extend single heliostat simulation tests

Refactored the test fixture to modularize heliostat setup, simulation, and result checking. Added helper methods for setting up different heliostat configurations, reading expected results, and comparing simulation outputs. Consolidated and parameterized test cases for various heliostat positions and configurations, improving code reuse and maintainability.

qualand committed:
Update single_heliostat_test.cpp

qualand committed:
Refactor heliostat field test and update efficiency checks

Removed per-heliostat ray count vectors in favor of total counters, updated ray counting logic, and adjusted test tolerances and efficiency calculations. Added gap settings for heliostats, included heat shield elements in simulation stages, and improved output information for test diagnostics. Commented out sun size checks and cleaned up test structure.

qualand committed:
Set number of threads in heliostat field test

Added a call to set the number of threads to 10 in the heliostat_field_test.cpp setup.

qualand committed:
removing unused files

qualand committed:
Replace std::exception with std::invalid_argument

taylorbrown75 commented:
Are we concerned about the test duration? Current develop takes ~6 minutes to run, while the windows debug with round robin tests is 46 minutes. Should we reduce the number of rays for the heliostat field tests, or wait to merge until the native runner multi threading is merged in?

qualand commented:
> Are we concerned about the test duration? Current develop takes ~6 minutes to run, while the windows debug with round robin tests is 46 minutes. Should we reduce the number of rays for the heliostat field tests, or wait to merge until the native runner multi threading is merged in?

The run times will be reduced once we merge the multi-threading native runner. We could skip the full field tests in debug configuration.

qualand committed:
Reduced the number of rays for full field tests

added multi-threading to the single heliostat tests

qualand committed:
removing full field tests in debug builds

Added a flag to include these.

qualand committed:
Merge branch 'develop' into round_robin_validation_tests

qualand committed:
Update heliostat_field_test.cpp

qualand committed:
increase tolerance due to multi-threaded seed number difference results

taylorbrown75 commented on .vscode/settings.json:
Should we ignore the .vscode folder for git?

qualand replied:
added.

taylorbrown75 commented on google-tests/unit-tests/simulation_data/cst-templates/single_heliostat_test.cpp:
```diff
@@ -0,0 +1,791 @@
+        EXPECT_NEAR(PeakFlux / 1.e3, expected_peak_flux, peak_tol * expected_peak_flux);
+
+        // RMS of flux values
+        if (!high_accuracy) calculate_receiver_flux_map(result, 100, 150, false);  // Re-calculate
```
Is this making the high accuracy flux map that is skipped on line 594? Can we call calculate_receiver_flux_map(result, 100, 150, false) once for either case?

qualand replied:
added comments for clarity.

taylorbrown75 commented on google-tests/unit-tests/simulation_data/cst-templates/heliostat_field_test.cpp:
```diff
@@ -0,0 +1,973 @@
+        }
+    }
+
+    void calculate_ray_counts(SimulationResult result) {
```
Recommend pass by reference since SimulationResult can be very large.

qualand replied:
addressed

taylorbrown75 commented on google-tests/unit-tests/simulation_data/cst-templates/heliostat_field_test.cpp:
```diff
@@ -0,0 +1,973 @@
+        NumberOfRays = 0;
+    }
+
+    bool calculate_receiver_flux_map(SimulationResult result, int nbinsx, int nbinsy, bool is_cylinder) {
```
Recommend pass by reference.

qualand replied:
addressed

taylorbrown75 commented on google-tests/unit-tests/simulation_data/cst-templates/heliostat_field_test.cpp:
```diff
@@ -0,0 +1,973 @@
+
+    }
+
+    void check_outputs(SimulationResult result) {
```
Recommend pass by reference.

qualand replied:
addressed

taylorbrown75 reviewed (approved):
Looks good, and will be very helpful for future validation. I added a few comments on places that might speed up the tests (with help from copilot).

qualand committed:
Merge branch 'develop' into round_robin_validation_tests

qualand committed:
Refactor heliostat field test for const reference and result saving

Updated several methods to take SimulationResult as a const reference. Enhanced simulate_check_outputs to optionally save results to files with descriptive filenames when save_results is true, replacing commented-out code with functional file output.

qualand committed:
Clarify flux map calculation comments in test

qualand committed:
Add .vscode/ to .gitignore

qualand committed:
Add gcov-ignore-parse-errors option to lcov command

qualand committed:
Update lcov command for coverage reporting

qualand committed:
Update coverage command to use profile update option

qualand committed:
Update coverage commands

qualand committed:
Update coverage commands for lcov in CI and CMake

qualand committed:
Update coverage settings

Removes redundant lcov ignore pattern in the GitHub Actions workflow and adds '-fprofile-update=atomic' to coverage build flags in CMake.

qualand committed:
Update coverage flags in CMakeLists for coretrace and tests

Removed '-fprofile-update=atomic' from coretrace coverage flags and added it to google-tests coverage link options.

qualand committed:
Disable several heliostat simulation tests

Commented out multiple test cases in heliostat_field_test.cpp and single_heliostat_test.cpp to temporarily disable them. Also updated coverage flags in CMakeLists.txt to use target_compile_options for -fprofile-update=atomic and target_link_options for --coverage. To address coverage issue.

qualand committed:
Enable coverage flags in CMake configuration

qualand committed:
Fix lcov command to handle coverage info errors

qualand committed:
Refactor coverage.yml for improved artifact management

Updated the coverage workflow to clean previous artifacts, adjust installation commands, and improve test execution.

qualand committed:
Update test CMakeLists for coverage and test inclusion

Removed atomic profile update compile option for coverage builds in google-tests/CMakeLists.txt. Adjusted CST test source inclusion logic to only add single_heliostat_test.cpp and heliostat_field_test.cpp for Release or FULL_FIELD_VALIDATION_TEST builds.

qualand committed:
Update lcov commands to ignore 'unused' errors

qualand committed:
Update lcov commands to handle mismatches

qualand committed:
Trying to fix coverage.yml

Removed mismatch ignore flag and profile-update=atomic"

qualand committed:
Trying profile-update=atomic for both C++ and C builds

qualand committed:
reverting atomic profile update

Removed coverage flags from CMake configuration.

qualand committed:
Enable round robin tests

Uncommented and activated previously disabled heliostat field and single heliostat simulation tests. Added save_results flag and logic to optionally save simulation outputs to CSV files. Minor formatting and cleanup in sun position test file.

qualand committed:
Add option to save ray data in heliostat field tests

Introduces a new 'save_raydata' flag to control saving of ray data to CSV files in heliostat_field_test.cpp. Ray data is now only saved if both 'save_results' and 'save_raydata' are true, providing more granular control over test output.

qualand committed:
Fixed task 1b focal lengths to canting bands

The original round-robin case setup for task 1b used the canting bands for determining the focal length rather than focal bands reported in the paper. This change reduced errors in result comparisons.
